File: Android/FitIn_v2/app/src/main/java/com/example/fitin_v2/ui/onboard/signin/SignInViewModel.java
package com.example.fitin_v2.ui.onboard.signin;

import android.app.Application;
import android.view.View;

import androidx.annotation.NonNull;
import androidx.lifecycle.AndroidViewModel;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;

import com.example.fitin_v2.model.AccountLoginDto;
import com.example.fitin_v2.repository.NewsRepository;
import com.example.fitin_v2.repository.UserRepository;

import io.reactivex.disposables.CompositeDisposable;

public class SignInViewModel extends AndroidViewModel {

    private UserRepository userRepository;
    private NewsRepository newsRepository;
    private final CompositeDisposable disposable = new CompositeDisposable();

    public MutableLiveData<String> email = new MutableLiveData<>();
    public MutableLiveData<String> password = new MutableLiveData<>();

    private MutableLiveData<AccountLoginDto> userMutableLiveData;

    private final MutableLiveData<Boolean> _eventSignIn = new MutableLiveData<Boolean>();
    LiveData<Boolean> eventSignIn;

    private final MutableLiveData<Boolean> _eventBack = new MutableLiveData<Boolean>();
    LiveData<Boolean> eventBack;

    public SignInViewModel(@NonNull Application application) {
        super(application);

        userRepository = new UserRepository(application);
        newsRepository = new NewsRepository(application);
    }

    public MutableLiveData<AccountLoginDto> getUser() {

        if (userMutableLiveData == null) {
            userMutableLiveData = new MutableLiveData<>();
        }
        return userMutableLiveData;
    }

    public LiveData<Boolean> getBack() {
        return eventBack = _eventBack;
    }

    public void onBack() {
        _eventBack.setValue(true);
    }

    public void onBackComplete() {
        _eventBack.setValue(false);
    }

    public LiveData<Boolean> getEventSignIn() {
        return eventSignIn = _eventSignIn;
    }

    public void onEventSignInComplete() {
        _eventSignIn.setValue(false);
    }


    public void onLogin(View view) {
        AccountLoginDto accountLoginDto = new AccountLoginDto(email.getValue(), password.getValue());
        userRepository.getToken(accountLoginDto);
        newsRepository.callNews();
        // 로그인 검증 설정하기 결과값 바탕으로 가져온 값에 따라서 홈으로 넘어갈지 말지 처리 추가하기
//        String validation = Preferences.getLogin("오류");
//        if(validation.equals("Success")) {
////            Log.e("token", "tokens " + token);
//            _eventSignIn.setValue(true);
//        } else if(validation.equals("Fail")) {
////            Log.e("token", "tokens" + token);
//            _eventSignIn.setValue(false);
//        } else {
//            _eventSignIn.setValue(false);
//        }
        // 뉴스 callAPI 데이터 불러와서 DB에 저장

        _eventSignIn.setValue(true);
    }

    @Override
    protected void onCleared() {
        super.onCleared();

        disposable.clear();
    }
}

File: Android/FitIn_Kotlin/app/src/main/java/com/example/fitin_kotlin/ui/onboard/signin/SignInViewModel.kt
package com.example.fitin_kotlin.ui.onboard.signin

import android.util.Log
import android.view.View
import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.example.fitin_kotlin.data.local.EncryptedSharedPreferenceController
import com.example.fitin_kotlin.data.model.network.request.RequestSignIn
import com.example.fitin_kotlin.data.repository.NewsRepository
import com.example.fitin_kotlin.data.repository.UserRepository
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.launch
import javax.inject.Inject

@HiltViewModel
class SignInViewModel @Inject constructor(
    private val userRepository: UserRepository,
    private val prefs: EncryptedSharedPreferenceController
) : ViewModel(){

    val email: MutableLiveData<String> = MutableLiveData<String>()
    val password: MutableLiveData<String> = MutableLiveData<String>()

    private val _eventSignIn = MutableLiveData<Boolean>()
    val eventSignIn: LiveData<Boolean>
        get() = _eventSignIn

    private val _errorMessage = MutableLiveData<String>()
    val errorMessage: LiveData<String>
        get() = _errorMessage

    fun onSignIn(view: View) {
        val requestSignIn = RequestSignIn(email.value, password.value)
        viewModelScope.launch {
            val signIn = userRepository.postSignIn(requestSignIn)
            when (signIn.isSuccessful) {
                true -> {
                    if (signIn.body()!!.state == 200) {
//                        Log.e("성공", signIn.body()!!.data.toString())
                        val result = signIn.body()!!.data.toString()
                        val accessToken = result.substring(result.indexOf("accessToken=") + 12, result.indexOf(", refreshToken"))
                        val refreshToken = result.substring(result.indexOf("refreshToken=") + 13, result.indexOf(", refreshTokenExpiresIn"))
                        prefs.setAccessToken(accessToken)
                        prefs.setRefreshToken(refreshToken)
//                        Log.e("Access Token", prefs.getAccessToken()!!)
//                        Log.e("Refresh Token", prefs.getRefreshToken()!!)
                        _eventSignIn.value = true
                    } else {
                        if (signIn.body()!!.message.isNotEmpty()) {
//                            Log.e("실패", signIn.body()!!.message)
                            val userNotExist = signIn.body()!!.message
                            _errorMessage.value = userNotExist
                        } else {
//                            Log.e("실패", signIn.body()!!.error[0].toString())
                            val result = signIn.body()!!.error[0].toString()
                            val failMessage = result.substring(result.lastIndexOf("=")+1,result.indexOf("}"))
//                            Log.e("안내 문자", failMessage)
                            _errorMessage.value = failMessage
                        }
                    }
                }
                else -> {
//                    Log.e("실패", "error " + signIn.message())
                    if (signIn.code() == 401) {
                        _errorMessage.value = "비밀번호가 일치하지 않습니다."
                    }
                    _eventSignIn.value = false
                }
            }
        }

    }

    fun onEventSignInComplete() {
        _eventSignIn.value = false
    }

    private val _eventKakaoSign = MutableLiveData<String?>()
    val eventKakaoSign: LiveData<String?>
        get() = _eventKakaoSign

    fun onEventKakao() {
        _eventKakaoSign.value = "KaKao"
    }

    fun onEventKakaoComplete() {
        _eventKakaoSign.value = null
    }

    private val _eventNaverSign = MutableLiveData<String?>()
    val eventNaverSign: LiveData<String?>
        get() = _eventNaverSign

    fun onEventNaver() {
        _eventNaverSign.value = "Naver"
    }

    fun onEventNaverComplete() {
        _eventNaverSign.value = null
    }

    private val _eventBack = MutableLiveData<Boolean>()
    val eventBack: LiveData<Boolean>
        get() = _eventBack

    fun onBack() {
        _eventBack.value = true
    }

    fun onBackComplete() {
        _eventBack.value = false
    }

    private val _eventFindId = MutableLiveData<Boolean>()
    val eventFindId: LiveData<Boolean>
        get() = _eventFindId

    fun onFindId() {
        _eventFindId.value = true
    }

    fun onFindIdComplete() {
        _eventFindId.value = false
    }

}